cmd/analyzetestresults/analyzetestresults.go

```diff
@@ -0,0 +1,221 @@
+}
+
+// setHeaderString initialises struct FailedTestCasesReport's
+// 'headerString' field based on phase at which PipelineRun failed
+func (f *FailedTestCasesReport) setHeaderString(overallJUnitSuites *reporters.JUnitTestSuites) {
+	if len(overallJUnitSuites.TestSuites) == 0 {
+		if f.clusterProvisionLog == "" && f.e2eTestLog == "" {
+			klog.Info("could not find any related artifacts")
+			f.failureType = OtherFailure
+			f.headerString = ":rotating_light: **Couldn't detect a specific failure, see the related PipelineRun for more details or consult with Konflux DevProd team.**\n"
+			return
+		}
+		if f.e2eTestLog != "" {
+			klog.Info("no JUnit file found - PipelineRun failed during running tests")
+			f.failureType = TestRunFailure
+			f.headerString = ":rotating_light: **No JUnit file found, see the log from running tests**: \n"
+			return
+		}
+		if f.clusterProvisionLog != "" {
+			klog.Info("failed to provision a cluster")
+			f.failureType = ClusterCreationFailure
+			f.headerString = ":rotating_light: **Failed to provision a cluster, see the log for more details**: \n"
+			return
+		}
+
+	} else {
+		klog.Info("The given PipelineRun failed while running tests")
+		f.failureType = TestCaseFailure
+		f.headerString = ":rotating_light: **Error occurred while running the E2E tests, list of failed Spec(s)**: \n"
+	}
+}
+
+// extractFailedTestCases initialises the FailedTestCasesReport struct's
+// 'failedTestCaseNames' field with the names of failed test cases
+// within given JUnitTestSuites -- if the given JUnitTestSuites is !nil.
+func (f *FailedTestCasesReport) extractFailedTestCases(overallJUnitSuites *reporters.JUnitTestSuites) {
+	switch f.failureType {
+	case OtherFailure:
+		return
+	case ClusterCreationFailure:
+		testCaseEntry := returnContentWrappedInDropdown(dropdownSummaryString, f.clusterProvisionLog)
+		f.failedTestCaseNames = append(f.failedTestCaseNames, testCaseEntry)
+		return
+	case TestRunFailure:
+		testCaseEntry := returnContentWrappedInDropdown(dropdownSummaryString, f.e2eTestLog)
+		f.failedTestCaseNames = append(f.failedTestCaseNames, testCaseEntry)
+		return
+	}
+	for _, testSuite := range overallJUnitSuites.TestSuites {
+		if testSuite.Failures > 0 || testSuite.Errors > 0 {
+			var tcMessage string
+			for _, tc := range testSuite.TestCases {
+				if tc.Failure != nil || tc.Error != nil {
+					klog.Infof("Found a Test Case (suiteName/testCaseName): %s/%s, that didn't pass", testSuite.Name, tc.Name)
+					switch {
+					case tc.Status == "timedout":
+						tcMessage = returnContentWrappedInDropdown(dropdownSummaryString, tc.SystemErr)
+					case tc.Failure != nil:
+						tcMessage = "```\n" + tc.Failure.Message + "\n```"
+					default:
+						tcMessage = "```\n" + tc.Error.Message + "\n```"
+					}
+
+					testCaseEntry := "* :arrow_right: " + "[**`" + tc.Status + "`**] " + tc.Name + "\n" + tcMessage
+					f.failedTestCaseNames = append(f.failedTestCaseNames, testCaseEntry)
+				}
+			}
+		}
+	}
+}
```

Comment:
I think we should move some of the nested `if` logic around checking for failure/errors greater than 0 and not Nil to the Report struct you have. 

Then all this function has to do is call a `getFailedTestCases` or something like that. That already returns a set of filtered failed test cases. 

Then this function becomes more of a formatter that iterates that list of already filtered test cases using that switch statement